packages/kbn-config-schema/src/types/object_type.test.ts

```diff
@@ -374,3 +374,120 @@ test('handles optional properties', () => {
+
+    expect(() => {
+      extended.validate({ initial: 'foo' });
+    }).toThrowErrorMatchingInlineSnapshot(
+      `"[added]: expected value of type [number] but got [undefined]"`
+    );
+
+    expect(() => {
+      extended.validate({ initial: 'foo', added: 42 });
+    }).not.toThrowError();
+
+    // asserting that the resulting type is valid
+    const value: TypeOf<typeof extended> = {
+      added: 12,
+      initial: 'foo',
+    };
+    expect(value).toBeDefined();
+  });
+
+  it('allows to extend an existing schema by removing properties', () => {
+    const origin = schema.object({
+      string: schema.string(),
+      number: schema.number(),
+    });
+
+    const extended = origin.extends({ number: undefined });
+
+    expect(() => {
+      extended.validate({ string: 'foo', number: 12 });
+    }).toThrowErrorMatchingInlineSnapshot(`"[number]: definition for this key is missing"`);
+
+    expect(() => {
+      extended.validate({ string: 'foo' });
+    }).not.toThrowError();
+
+    // asserting that the resulting type is valid
+    const value: TypeOf<typeof extended> = {
+      string: 'foo',
+    };
+    expect(value).toBeDefined();
+  });
+
+  it('allows to extend an existing schema by overriding an existing properties', () => {
+    const origin = schema.object({
+      string: schema.string(),
+      mutated: schema.number(),
+    });
+
+    const extended = origin.extends({
+      mutated: schema.string(),
+    });
+
+    expect(() => {
+      extended.validate({ string: 'foo', mutated: 12 });
+    }).toThrowErrorMatchingInlineSnapshot(
+      `"[mutated]: expected value of type [string] but got [number]"`
+    );
+
+    expect(() => {
+      extended.validate({ string: 'foo', mutated: 'bar' });
+    }).not.toThrowError();
+
+    // asserting that the resulting type is valid
+    const value: TypeOf<typeof extended> = {
+      string: 'foo',
+      mutated: 'bar',
+    };
+    expect(value).toBeDefined();
+  });
+
+  it('properly infer the type from optional properties', () => {
+    const origin = schema.object({
+      original: schema.maybe(schema.string()),
+      mutated: schema.maybe(schema.number()),
+      removed: schema.maybe(schema.string()),
+    });
+
+    const extended = origin.extends({
+      removed: undefined,
+      mutated: schema.string(),
+    });
+
+    expect(() => {
+      extended.validate({ original: 'foo' });
+    }).toThrowErrorMatchingInlineSnapshot(
+      `"[mutated]: expected value of type [string] but got [undefined]"`
+    );
+    expect(() => {
+      extended.validate({ original: 'foo' });
+    }).toThrowErrorMatchingInlineSnapshot(
+      `"[mutated]: expected value of type [string] but got [undefined]"`
+    );
+    expect(() => {
+      extended.validate({ original: 'foo', mutated: 'bar' });
+    }).not.toThrowError();
+
+    // asserting that the resulting type is valid
+    let value: TypeOf<typeof extended> = {
+      original: 'foo',
+      mutated: 'bar',
+    };
+    value = {
```

Comment:
can be deleted?

Reply:
Yea, I really don't like there TS typecheck validation 'assertions' in tests, and would really love to have a better option

I see we do have some `tsd` tests in 

https://github.com/elastic/kibana/blob/6436862dc52942a060246af30e63e0b4d2db17bf/packages/kbn-utility-types/test-d/union_to_intersection.ts#L20-L28

It would probably be way better to use this, wdyt?

Reply:
I'm fine to keep it as is and wait for https://github.com/elastic/kibana/issues/53762. Adding tool requires a lot of additional work: reach a consensus on a test tool, update infrastructure to run a new type of tests, make sure test reporter works with a test tool, etc.

Reply:
we can't properly use `expectError` until TS 3.9 for the `@expect-error` annotation, but the `expectType` use is pretty straightforward (and already used by `kbn-utility-types`): https://github.com/elastic/kibana/pull/68067/commits/de6db73c14586d14926eeb025ad617ea14c9378e

That was mostly a test, I can revert the commit if we want.

Reply:
Alright, I thought you wanted to extract them out of the tests. I'm fine to use `tsd` in jest tests.